app/upgrades/v17/upgrades.go

```diff
@@ -32,6 +34,51 @@ func CreateUpgradeHandler(
 			CosmwasmWeight:     300,
 		})
 
+		// get all the existing CL pools
+		pools, err := keepers.ConcentratedLiquidityKeeper.GetPools(ctx)
+		if err != nil {
+			return nil, err
+		}
+
+		for _, pool := range pools {
+			// migrate twap records for CL Pools
+			err = FlipTwapSpotPriceRecords(ctx, []uint64{pool.GetId()}, keepers)
+			if err != nil {
+				return nil, err
+			}
+		}
+
 		return migrations, nil
 	}
 }
+
+func FlipTwapSpotPriceRecords(ctx sdk.Context, poolIds []uint64, keepers *keepers.AppKeepers) error {
+	for _, poolId := range poolIds {
+		// check that this is a cl pool
+		_, err := keepers.ConcentratedLiquidityKeeper.GetConcentratedPoolById(ctx, poolId)
+		if err != nil {
+			return err
+		}
+
+		// check that the twap record exists
+		clPoolTwapRecords, err := keepers.TwapKeeper.GetAllMostRecentRecordsForPool(ctx, poolId)
+		if err != nil {
+			return err
+		}
```

Comment:
So we call this in the upgrade handler right after we create a bunch of CL pools. Would these pools return an error here? Or when we create the pool do they have a TWAP record and this would not error?

If the former, we should check the error and if the error implies there is no TWAP record we might want to just continue.

Reply:
nope no error afaik, once pool gets created we create an empty TWAP record ref: https://github.com/osmosis-labs/osmosis/blob/main/x/twap/logic.go#L83

i also tested this and this is what it looks like; 
```
[{3 atom uosmo 1 2023-08-02 19:28:24.422228358 +0000 UTC 0.000000000000000000 0.000000000000000000 0.000000000000000000 0.000000000000000000 0.000000000000000000 2023-08-02 19:28:24.422228358 +0000 UTC}]
```